File: ts/797.ts
function allPathsSourceTarget(graph: number[][]): number[][] {
    const n = graph.length;
    const res: number[][] = [];

    const dfs = (path: number[], node: number): void => {
        if (node === n - 1) {
            res.push(path.slice());
            return;
        }

        for (const next of graph[node]) {
            path.push(next);
            dfs(path, next);
            path.pop();
        }
    }

    dfs([0], 0);

    return res;
};

File: kt/797.kt
class Solution {
    fun allPathsSourceTarget(graph: Array<IntArray>): List<List<Int>> {
        val n = graph.size
        val ans = mutableListOf<List<Int>>()
        fun dfs(path: MutableList<Int>, node: Int) {
            if (node == n - 1) {
                ans.add(path.toList())
                return
            }
            
            for (next in graph[node]) {
                path.add(next)
                dfs(path, next)
                path.removeAt(path.size - 1)
            }
        }
        
        dfs(mutableListOf(0), 0)
        return ans
    }
}